### PR TITLE
New VESTEC filters

### PR DESCRIPTION
**Pull request by pierre-guillou**
This PR adds new utility filters in relation with the VESTEC project:
* PersistenceDiagramDistanceMatrix extracts the distance matrix code from the PersistenceDiagramClustering module (code has been removed from the latter module)
* StringArrayConverter generates integer arrays from string arrays
* PointSetToCurve creates a (optionally closable) curve from a point set by linking consecutive points with a line
* LDistanceMatrix generates a distance matrix from a VTI using the LDistance metric
* MatrixToHeatMap creates a VTU heat map from a vtkTable distance matrix

Previously discussed std::pow optimizations will occur in a later PR.

Enjoy!
Pierre

### CODE REVIEW

**Comment by julien-tierny** [comment]
hi pierre,

thanks a lot, that's awesome.
could you have a look at fixing the ci though?
@CharlesGueunet may help.

cheers,
-- 
Dr Julien Tierny
CNRS Researcher
Sorbonne Universite
http://lip6.fr/Julien.Tierny

**Comment by pierre-guillou** [comment]
Seems to me that Boost is not correctly built during the "Build boost" phase. I can't see the result of the .\b2 command in the output log.

File: core/vtk/ttkLDistanceMatrix/ttkLDistanceMatrix.cpp [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+  for(size_t i = 0; i < nInputs; ++i) {
+    inputData[i] = vtkImageData::SafeDownCast(blocks->GetBlock(i));
+  }
+
+  // sanity check: data non null and same data size
+  {
+    std::set<vtkIdType> sizes{};
+    for(size_t i = 0; i < nInputs; ++i) {
+      if(inputData[i] == nullptr) {
+        this->printErr("Some input block is not a vtkImageData");
```
Comment (julien-tierny) [inline core/vtk/ttkLDistanceMatrix/ttkLDistanceMatrix.cpp]: are you restricting to vtkImageData?
(this should be generic vtkDataSet)

File: core/vtk/ttkLDistanceMatrix/ttkLDistanceMatrix.cpp [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+      return 0;
+    }
+  }
+
+  // Get output
+  auto DistTable = vtkTable::GetData(outputVector);
+
+  std::vector<std::vector<double>> distMatrix(nInputs);
+
+  LDistance worker{};
```
Comment (julien-tierny) [inline core/vtk/ttkLDistanceMatrix/ttkLDistanceMatrix.cpp]: please create a new class in core/base/ for this code (lDistanceMatrix).

File: core/vtk/ttkPointSetToCurve/ttkPointSetToCurve.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
+/// \class ttkPointSetToCurve
+/// \ingroup vtk
+/// \author Jonas Lukasczyk <jl@jluk.de>
```
Comment (julien-tierny) [inline core/vtk/ttkPointSetToCurve/ttkPointSetToCurve.h]: you might want to put your own contact info instead.

File: core/vtk/ttkStringArrayConverter/ttkStringArrayConverter.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
+/// \class ttkStringArrayConverter
+/// \ingroup vtk
+/// \author Jonas Lukasczyk <jl@jluk.de>
```
Comment (julien-tierny) [inline core/vtk/ttkStringArrayConverter/ttkStringArrayConverter.h]: same here.

File: paraview/PointSetToCurve/PointSetToCurve.xml [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        <Documentation>
+          Close the curve by generating a line cell between the last
+          point and the first point of the point set.
+        </Documentation>
+      </IntVectorProperty>
+
+      <PropertyGroup panel_widget="Line" label="Input options">
+        <Property name="InputOrderingArray" />
+      </PropertyGroup>
+
```
Comment (julien-tierny) [inline paraview/PointSetToCurve/PointSetToCurve.xml:73]: no   ${DEBUG_WIDGETS}?

File: paraview/StringArrayConverter/StringArrayConverter.xml [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        </ArrayListDomain>
+        <Documentation>
+          Select the input string array.
+        </Documentation>
+      </StringVectorProperty>
+
+      <PropertyGroup panel_widget="Line" label="Input options">
+        <Property name="InputStringArray" />
+      </PropertyGroup>
+
```
Comment (julien-tierny) [inline paraview/StringArrayConverter/StringArrayConverter.xml:59]: no   ${DEBUG_WIDGETS}?

File: paraview/MatrixToHeatMap/MatrixToHeatMap.xml [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            matching ones will be selected.
+         </Documentation>
+      </StringVectorProperty>
+
+      <PropertyGroup panel_widget="Line" label="Input options">
+        <Property name="SelectFieldsWithRegexp" />
+        <Property name="ScalarFields" />
+        <Property name="Regexp" />
+      </PropertyGroup>
+
```
Comment (julien-tierny) [inline paraview/MatrixToHeatMap/MatrixToHeatMap.xml:94]: no   ${DEBUG_WIDGETS}?

File: paraview/PersistenceDiagramClustering/PersistenceDiagramClustering.xml [hunk 8]
```diff
@@ -490,8 +428,6 @@ between matched critical points.
       <OutputPort name="Clustered Diagrams" index="0" id="port0" />
       <OutputPort name="Cluster Centroids" index="1" id="port1" />
       <OutputPort name="Matchings" index="2" id="port2" />
-      <OutputPort name="Diagrams Distance" index="3" id="port3" />
-
 
       <Hints>
```
Comment (julien-tierny) [inline paraview/PersistenceDiagramClustering/PersistenceDiagramClustering.xml:382]: no   ${DEBUG_WIDGETS}?

Comment (pierre-guillou) [reply paraview/PersistenceDiagramClustering/PersistenceDiagramClustering.xml:382]: There is one now!

File: paraview/StringArrayConverter/StringArrayConverter.xml (hunk 6 above)
Comment (pierre-guillou) [reply paraview/StringArrayConverter/StringArrayConverter.xml:59]: I felt there was no need here: no huge computation needing parallelism, no much log.

File: paraview/MatrixToHeatMap/MatrixToHeatMap.xml (hunk 7 above)
Comment (pierre-guillou) [reply paraview/MatrixToHeatMap/MatrixToHeatMap.xml:94]: Same as before, a simple sequential computation, no much log.

File: core/vtk/ttkLDistanceMatrix/ttkLDistanceMatrix.cpp (hunk 1 above)
Comment (pierre-guillou) [reply core/vtk/ttkLDistanceMatrix/ttkLDistanceMatrix.cpp]: I've updated the code to be more generic. vtkDataSet are now supported here.

File: core/vtk/ttkLDistanceMatrix/ttkLDistanceMatrix.cpp (hunk 2 above)
Comment (pierre-guillou) [reply core/vtk/ttkLDistanceMatrix/ttkLDistanceMatrix.cpp]: Done! There is now a LDistanceMatrix base module.

File: core/vtk/ttkPointSetToCurve/ttkPointSetToCurve.h (hunk 3 above)
Comment (pierre-guillou) [reply core/vtk/ttkPointSetToCurve/ttkPointSetToCurve.h]: Fixed now

File: core/vtk/ttkStringArrayConverter/ttkStringArrayConverter.h (hunk 4 above)
Comment (pierre-guillou) [reply core/vtk/ttkStringArrayConverter/ttkStringArrayConverter.h]: Also fixed

File: paraview/PointSetToCurve/PointSetToCurve.xml (hunk 5 above)
Comment (pierre-guillou) [reply paraview/PointSetToCurve/PointSetToCurve.xml:73]: Same as before, no real need here.

**Comment by julien-tierny** [comment]
thanks for synchronizing that PR with the current dev branch (azure fix).
-- 
Dr Julien Tierny
CNRS Researcher
Sorbonne Universite
http://lip6.fr/Julien.Tierny

File: paraview/MatrixToHeatMap/MatrixToHeatMap.xml [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+      ${DEBUG_WIDGETS}
+
+      <Hints>
+        <ShowInMenu category="TTK - High Dimension / Point Cloud Data" />
```
Comment (julien-tierny) [inline paraview/MatrixToHeatMap/MatrixToHeatMap.xml:98]: no so sure about this category (but this is fixable after the merge)

**Comment by julien-tierny** [comment]
alright pierre, this is some great work. 
thanks a lot!

I'm no so sure about the pv category for DistanceMatrixToHeatMap but we can arrange that differently later.

thanks!
-- 
Dr Julien Tierny
CNRS Researcher
Sorbonne Universite
http://lip6.fr/Julien.Tierny

**Pull request merged**